nessie/jobs/sync_canvas_requests_snapshots.py

```diff
@@ -0,0 +1,111 @@
+from nessie.lib import metadata
+from nessie.lib.dispatcher import dispatch
+
+"""Logic for Canvas snapshot sync job."""
+
+
+def delete_objects_with_prefix(prefix, whitelist=[]):
+    keys_to_delete = []
+    existing_keys = s3.get_keys_with_prefix(prefix)
+    if existing_keys is None:
+        app.logger.error('Error listing keys, aborting job.')
+        return False
+    for key in existing_keys:
+        filename = key.split('/')[-1]
+        if filename not in whitelist:
+            keys_to_delete.append(key)
+    app.logger.info(
+        f'Found {len(existing_keys)} key(s) matching prefix "{prefix}", {len(existing_keys) - len(keys_to_delete)} '
+        f'key(s) in whitelist, will delete {len(keys_to_delete)} object(s)')
+    if not keys_to_delete:
+        return True
+    if s3.delete_objects(keys_to_delete):
+        metadata.delete_canvas_snapshots(keys_to_delete)
+        return True
+    else:
+        return False
+
+
+class SyncCanvasRequestsSnapshots(BackgroundJob):
+
+    @classmethod
+    def generate_job_id(cls):
+        return 'sync_canvas_requests_' + str(int(time.time()))
+
+    def run(self, cleanup=True):
+        job_id = self.generate_job_id()
+        app.logger.info(f'Starting Canvas snapshot sync job... (id={job_id})')
+
+        snapshot_response = canvas_data.get_snapshots()
+        if not snapshot_response:
+            raise BackgroundJobError('Error retrieving Canvas data snapshots, aborting job.')
+        snapshots = snapshot_response.get('files', [])
+
+        def should_sync(snapshot):
+            return snapshot['table'] == 'requests' and snapshot['partial'] is False
+
+        snapshots_to_sync = [s for s in snapshots if should_sync(s)]
+        app.logger.info(f'Will sync {len(snapshots_to_sync)} of {len(snapshots)} available files from Canvas Data.')
+
+        success = 0
+        failure = 0
+
+        for snapshot in snapshots_to_sync:
+            metadata.create_canvas_sync_status(
+                job_id=job_id,
+                filename=snapshot['filename'],
+                canvas_table=snapshot['table'],
+                source_url=snapshot['url'],
+            )
+            if snapshot['table'] == 'requests':
+                key_components = [app.config['LOCH_S3_CANVAS_DATA_PATH_HISTORICAL'], snapshot['table'], snapshot['filename']]
+
+            key = '/'.join(key_components)
+            response = dispatch('sync_file_to_s3', data={'canvas_sync_job_id': job_id, 'url': snapshot['url'], 'key': key})
+
+            if not response:
+                app.logger.error('Failed to dispatch S3 sync of snapshot ' + snapshot['filename'])
+                metadata.update_canvas_sync_status(job_id, key, 'error', details=f'Failed to dispatch: {response}')
+                failure += 1
+            else:
+                app.logger.info('Dispatched S3 sync of snapshot ' + snapshot['filename'])
+                success += 1
+
+        if cleanup:
```

Comment:
@sandeepmjay If the job is for archival purposes, will it need to make use of this cleanup routine?

Reply:
It might if canvas changes the list of files in their inventory. It is more a fail-safe so we don't house redundant data files.